### ISSUE TITLE
Implement push notifications 

### ISSUE THREAD

**Issue by StephenRedd** [NullDesk/TicketDesk#40]
This mechanism must live, at least partially, in-process. It will need to leverage the razor view engine to generate email notes, if not other types. Since it lives in-process, it should probably be triggered by the process that created the event notification entry in the DB, not by monitoring the event notificaitons in the DB --similar to how DbContext signals that search items should be added to the search index.

* Monitor event notifications marked for push
* Evaluate if push notes should be sent
* Generate the push note content
* Queue push note content for delivery

**Comment by brunoshine** [comment]
Hi,

I've built a notifications handler that could probably help you out on this (http://goo.gl/klWIZs). It abstracts the notifications mechanism, delegating that task to a list of event subscribers. It's based on MEF to discover the handlers and register them.

Let me know know your thoughts on this :)

Cheers

**Comment by StephenRedd** [comment]
I appreciate the link, and the code looks great! I'm about half-way through the notifications module for TD 2.5, much of which is on the develop branch already. The rest will follow later this week. So far, the part that deals with multiple delivery providers is shaping up to be very similar to your design. The only significant difference being that TD 2.5 uses SimpleInjector instead of MEF. 

I [wrote a bit about MEF and TicketDesk](http://www.reddnet.net/ticketdesk-3-dev-diary-mef-ioc-and-architectural-design/) a while back when I was brainstorming for TD3. I have a lot of respect for MEF as an extensibility framework, but the notification delivery providers are one of the few of TicketDesk's requirements that falls squarely in the realm of extensibility. The rest are more aligned to traditional IoC/DI concerns, so I went with SimpleInjector instead. 

But the lines between extensibility and IoC/DI frameworks is very thin, and in most cases either type can handle both areas of concern without too much trouble. In either case, the overall design remains largely identical. 

Outside of the delivery providers though, the rest of TicketDesk's notification system is rather complex; much more so than I'd expected. Mostly that comes from the fact that only a small queueing part runs in the main web process where it has awareness of both the web context, and the core domain context. 

The monitoring, scheduling, anti-noise system, and delivery all have to be able to operate in a completely external process (a WebJob, windows service, worker-role, etc.). While this part can be a consumer of the domain context, it has to be able to run entirely without access to the web and identity contexts. This has turned out to be quite an interesting challenge.

One thing that will be missing from the initial beta of TD 2.5 is full support for runtime instantiation of plug-in modules. Currently, the modules do have to be referenced at compile time. I have prototyped a solution for that using SimpleInjector, and it should be ready in time for the final release.
